clang/test/SemaCXX/diagnose_if-warning-group.cpp

```diff
@@ -0,0 +1,45 @@
+// RUN: %clang_cc1 %s -verify -fno-builtin -Werror=comment -Wno-error=abi -Wfatal-errors=assume -Wno-fatal-errors=assume
+
+#define diagnose_if(...) __attribute__((diagnose_if(__VA_ARGS__)))
+
+template <bool b>
+void diagnose_if_wcomma() diagnose_if(b, "oh no", "warning", "comma") {}
+
+template <bool b>
+void diagnose_if_wcomment() diagnose_if(b, "oh no", "warning", "comment") {}
+
+void bougus_warning() diagnose_if(true, "oh no", "warning", "bougus warning") {} // expected-error {{unknown warning group}}
+
+void show_in_system_header() diagnose_if(true, "oh no", "warning", "assume", "Banane") {} // expected-error {{'diagnose_if' attribute takes no more than 4 arguments}}
+
+
+void diagnose_if_wabi_default_error() diagnose_if(true, "ABI stuff", "error", "abi") {}
+void diagnose_assume() diagnose_if(true, "Assume diagnostic", "warning", "assume") {}
+
+void call() {
+  diagnose_if_wcomma<true>(); // expected-warning {{oh no}}
+  diagnose_if_wcomma<false>();
+  diagnose_if_wcomment<true>(); // expected-error {{oh no}}
+  diagnose_if_wcomment<false>();
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wcomma"
+  diagnose_if_wcomma<true>();
+  diagnose_if_wcomment<true>(); // expected-error {{oh no}}
+#pragma clang diagnostic pop
+
+#pragma clang diagnostic push
+#pragma clang diagnostic ignored "-Wcomment"
+  diagnose_if_wcomma<true>(); // expected-warning {{oh no}}
+  diagnose_if_wcomment<true>();
+#pragma clang diagnostic pop
+
+  diagnose_if_wcomma<true>(); // expected-warning {{oh no}}
+  diagnose_if_wcomment<true>(); // expected-error {{oh no}}
+
+  diagnose_if_wabi_default_error(); // expected-warning {{ABI stuff}}
```

Comment:
Do we want to allow `-Wno-error=foo` to work for diagnostics which are explicitly an error in the attribute? We don't let users do things like that for other errors and that suggests that `"error"` in `disable_if` is actually "warning that defaults to an error" rather than an error.

Reply:
If you don't give the attribute a warning group, it should still produce an error. With a warning group it's currently implemented as a warning that defaults to an error, since it doesn't seem to make a ton of sense to add a warning group but then ask for it to never be a warning.

Reply:
Ah, so the idea is:
```
diagnose_if(true, "oh no", "error" "thing") // #1
diagnose_if(true, "oh no", "error", "") // #2
diagnose_if(true, "oh no", "error") // #3
```
#1 is a warning that default to an error, #2 and #3 are errors with no ability to disable?

Reply:
I think #2 is currently rejected, but yeah, that's the idea.

Reply:
okay, that sounds reasonable to me (and I don't have strong opinions about supporting #2), thanks!